File: src/main/java/domain/klondike/SmallHarp.java
package domain.klondike;

import domain.Solitaire;
import domain.SolitaireContainerTypes;
import domain.ui.CalculatedPlacement;
import domain.ui.HorizontalPlacement;
import domain.ui.Layout;
import domain.ui.VerticalPlacement;

import java.awt.*;

/**
 * rules are same. Only layout is different
 */
public class SmallHarp extends KlondikeDomain {

    private Layout layout;

    @Override
    public Layout getLayout() {
        if (layout == null) {
            Point[] anchors = new Point[7];
            for (int i = 0; i < 7; i++) {
                anchors[i] = new Point(520 - 80*i, 40);
            }

            layout = new Layout()
                .add(SolitaireContainerTypes.Tableau,
                    new CalculatedPlacement(anchors, Solitaire.card_width, Solitaire.card_height+13*Solitaire.card_overlap))
                .add(SolitaireContainerTypes.Foundation,
                    new VerticalPlacement(new Point (600, 40), Solitaire.card_width, Solitaire.card_height, Solitaire.card_gap))
                .add(SolitaireContainerTypes.Stock, new HorizontalPlacement(new Point(580, 40 + 4*(Solitaire.card_height + Solitaire.card_gap)),
                    Solitaire.card_width,  Solitaire.card_height,  Solitaire.card_gap))
                .add(SolitaireContainerTypes.Waste, new HorizontalPlacement(new Point(500, 40 + 4*(Solitaire.card_height + Solitaire.card_gap)),
                    Solitaire.card_width,  Solitaire.card_height,  Solitaire.card_gap));
        }

        return layout;
    }

    public SmallHarp() {
        super("SmallHarp");
    }
}

File: src/main/scala/org/combinators/solitaire/klondike/smallharp.scala
package org.combinators.solitaire.klondike

import org.combinators.solitaire.domain._

package object smallharp extends variationPoints {

  override val klondikeLayout:Map[ContainerType, Seq[Widget]] = Map (
    Foundation -> horizontalPlacement(10, 10, 4, card_height),
    StockContainer -> horizontalPlacement(440, 10, 1, card_height),
    Tableau -> horizontalPlacement(10, 200, num = 7, height = card_height*8),  // estimate
    Waste -> horizontalPlacement(460 + card_width, 10, 1, card_height),
  )

  def smallHarpDeal: Seq[DealStep] = {
    var deal:Seq[DealStep] = Seq()

    // each of the BuildablePiles gets a number of facedown cards, 0 to first Pile, 1 to second pile, etc...
    // don't forget zero-based indexing.
    for (pileNum <- 0 until 6) {
      deal = deal :+ DealStep(ElementTarget(Tableau, pileNum), new Payload(false, 6-pileNum))
    }

    // finally each one gets a single faceup Card, and deal one to waste pile
    //add(new DealStep(new ContainerTarget(SolitaireContainerTypes.Tableau), new Payload()));
    deal = deal :+ DealStep(ContainerTarget(Tableau))

    // finally to deal cards
    deal = deal :+ DealStep(ContainerTarget(Waste), Payload(true, numToDeal))
    deal
  }

  val smallharp:Solitaire = {

    Solitaire( name="SmallHarp",
      structure = klondikeMap,
      layout = Layout(klondikeLayout),
      deal = smallHarpDeal,

      specializedElements = Seq(WastePile),

      moves = Seq(tableauToTableau, wasteToTableau, tableauToFoundation, wasteToFoundation, deckDeal, deckResetFromWaste, flipTableau),

      // fix winning logic
      logic = BoardState(Map(Foundation -> 52)),
      customizedSetup = Seq.empty
    )
  }
}
